File: scripts/algorithms/E/Even Odd Tree/Even Odd Tree.py
from collections import deque
# Runtime:838ms 44.05% || Memory: 40.8mb 57.22%
# O(n) || O(h); where h is the height of the tree

class Solution:
    def isEvenOddTree(self, root: Optional[TreeNode]) -> bool:
        if not root:
            return False

        level = 0
        evenOddLevel = {0:1, 1:0}
        queue = deque([root])

        while queue:
            prev = 0
            for _ in range(len(queue)):
                currNode = queue.popleft()
                comparison = {0:prev < currNode.val, 1:prev > currNode.val}
                if currNode.val % 2 != evenOddLevel[level % 2]:
                    return False
                else:
                    if prev != 0 and comparison[level % 2]:
                        prev = currNode.val
                    elif prev == 0:
                        prev = currNode.val
                    else:
                        return False

                if currNode.left:
                    queue.append(currNode.left)

                if currNode.right:
                    queue.append(currNode.right)

            level += 1

        return True

File: scripts/algorithms/E/Even Odd Tree/Even Odd Tree.js
// Runtime: 172 ms (Top 86.4%) | Memory: 104.29 MB (Top 45.7%)

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
var isEvenOddTree = function(root) {
    let queue = [root];
    let steps = 0;

    while(queue.length>0){
        const currLength = queue.length;
        let nextQueue=[];

        for(let i=0; i<currLength;i++){
            const node = queue[i];
            if(steps%2 !== 0){
                if((i<currLength-1 && node.val<=queue[i+1].val) || node.val%2 !== 0){
                    return false;
                }
            }else if((i<currLength-1 && node.val>=queue[i+1].val)|| node.val%2 === 0){
                return false;
            }
            if(node.left) nextQueue.push(node.left);
            if(node.right) nextQueue.push(node.right);
        }
        queue = nextQueue;
        steps++;
    }

    return true;
};